k-distribution/k-tutorial/1_basic/03_parsing/README.md

```diff
@@ -1 +1,627 @@
+when we parse this program, we see the following ambiguity as an error message:
+
+```
+[Error] Inner Parser: Parsing ambiguity.
+1: syntax Stmt ::= "if" "(" Exp ")" Stmt
+
+`if(_)__LESSON-03-I_Stmt_Exp_Stmt`(`true_LESSON-03-I_Exp`(.KList),`if(_)_else__LESSON-03-I_Stmt_Exp_Stmt_Stmt`(`false_LESSON-03-I_Exp`(.KList),`;_LESSON-03-I_Stmt`(.KList),`;_LESSON-03-I_Stmt`(.KList)))
+2: syntax Stmt ::= "if" "(" Exp ")" Stmt "else" Stmt
+
+`if(_)_else__LESSON-03-I_Stmt_Exp_Stmt_Stmt`(`true_LESSON-03-I_Exp`(.KList),`if(_)__LESSON-03-I_Stmt_Exp_Stmt`(`false_LESSON-03-I_Exp`(.KList),`;_LESSON-03-I_Stmt`(.KList)),`;_LESSON-03-I_Stmt`(.KList))
+        Source(./dangling-else.if)
+        Location(1,1,1,30)
+```
+
+Roughly, we see that the ambiguity is between an `if` with an `else` or an `if`
+without an `else`. Siknce we want to pick the first parse, we can tell K to
+"avoid" the second parse with the `avoid` attribute. Consider the following
+modified definition (`lesson-03-j.k`):
+
+```k
+module LESSON-03-I
+
+  syntax Exp ::= "true" | "false"
+  syntax Stmt ::= "if" "(" Exp ")" Stmt
+                | "if" "(" Exp ")" Stmt "else" Stmt [avoid]
+                | ";"
+endmodule
+```
+
+Here we have added the `avoid` attribute to the `else` production. As a result,
+when an ambiguity occurs and one or more of the possible parses has that symbol
+at the top of the ambiguous part of the parse, we remove those parses from
+consideration and consider only those remaining. The `prefer` attribute behaves
+similarly, but instead removes all parses which do not have that attribute.
+In both cases, no action is taken if the parse is not ambiguous.
+
+## Tokens
+
+So far we have seen how we can define the grammar of a language. However,
+the grammar is not the only relevant part of parsing a language. Also relevant
+is the lexical syntax of the language. Thus far, we have implicitly been using
+K's automatic lexer generation to generate a token in the scanner for each
+terminal in our grammar. However, sometimes we wish to define more complex
+lexical syntax. For example, consider the case of integers in C: an integer
+consists of a decimal, octal, or hexadecimal number followed by an optional
+suffix indicating the type of the literal.
+
+In theory it would be possible to define this syntax via a grammar, but not
+only would it be cumbersome and tedious, you would also then have to deal with
+an AST generated for the literal which is not convenient to work with.
+
+Instead of doing this, K allows you to define **token** productions, where
+a production consists of a regular expression followed by the `token`
+attribute, and the resulting AST consists of a typed string containing the
+value recognized by the regular expression.
+
+For example, we might define the integer token in C with the following
+production:
+
+```k
+syntax IntConstant ::= r"(([1-9][0-9]*)|(0[0-7]*)|(0[xX][0-9a-fA-F]+))(([uU][lL]?)|([uU]((ll)|(LL)))|([lL][uU]?)|(((ll)|(LL))[uU]?))?" [token]
+```
```

Comment:
This is much too overwhelming of an example to start with. Let's begin with just something like `[1-9][0-9]*` or the like, and then show them this one.